File: ios/Classes/BaiduMobStatPlugin.m
#import "BaiduMobStatPlugin.h"
#import "BaiduMobStat.h"

@implementation BaiduMobStatPlugin
+ (void)registerWithRegistrar:(NSObject<FlutterPluginRegistrar> *)registrar {
    FlutterMethodChannel *channel = [FlutterMethodChannel
                                     methodChannelWithName:@"fl_baidu_mob_stat"
                                     binaryMessenger:[registrar messenger]];
    BaiduMobStatPlugin *instance = [[BaiduMobStatPlugin alloc] init];
    [registrar addMethodCallDelegate:instance channel:channel];
}

- (void)handleMethodCall:(FlutterMethodCall *)call result:(FlutterResult)result {
    if ([call.method isEqualToString:@"setApiKey"]) {
        [[BaiduMobStat defaultStat] startWithAppId:call.arguments];
        [BaiduMobStat defaultStat].platformType = 2;
        result(@(YES));
    } else if ([call.method isEqualToString:@"setAppVersionName"]) {
        [[BaiduMobStat defaultStat] setShortAppVersion:call.arguments];
        result(@(YES));
    } else if ([call.method isEqualToString:@"setAppChannel"]) {
        [[BaiduMobStat defaultStat] setChannelId:call.arguments];
        result(@(YES));
    } else if ([call.method isEqualToString:@"setDebug"]) {
        [BaiduMobStat defaultStat].enableDebugOn = [call.arguments boolValue];
        result(@(YES));
    } else if ([call.method isEqualToString:@"logEvent"]) {
        NSString *eventId = call.arguments[@"eventId"];
        NSDictionary *attributes = [self validArgument:call.arguments[@"attributes"]];
        [[BaiduMobStat defaultStat] logEvent:eventId attributes:attributes];
        result(@(YES));
    } else if ([call.method isEqualToString:@"logDurationEvent"]) {
        NSString *eventId = call.arguments[@"eventId"];
        NSString *eventLabel = call.arguments[@"label"];
        NSInteger duration = [call.arguments[@"duration"] integerValue];
        NSDictionary *attributes = [self validArgument:call.arguments[@"attributes"]];
        [[BaiduMobStat defaultStat] logEventWithDurationTime:eventId eventLabel:eventLabel durationTime:duration attributes:attributes];
        result(@(YES));
    } else if ([call.method isEqualToString:@"eventStart"]) {
        [[BaiduMobStat defaultStat] eventStart:call.arguments[@"eventId"] eventLabel:call.arguments[@"label"]];
        result(@(YES));
    } else if ([call.method isEqualToString:@"eventEnd"]) {
        NSString *eventId = call.arguments[@"eventId"];
        NSString *label = call.arguments[@"label"];
        NSDictionary *attributes = [self validArgument:call.arguments[@"attributes"]];
        [[BaiduMobStat defaultStat] eventEnd:eventId eventLabel:label attributes:attributes];
        result(@(YES));
    } else if ([call.method isEqualToString:@"pageStart"]) {
        [[BaiduMobStat defaultStat] pageviewStartWithName:call.arguments];
        result(@(YES));
    } else if ([call.method isEqualToString:@"pageEnd"]) {
        [[BaiduMobStat defaultStat] pageviewEndWithName:call.arguments];
        result(@(YES));
    } else if ([call.method isEqualToString:@"getDeviceCuId"]) {
        result([[BaiduMobStat defaultStat] getDeviceCuid]);
    } else if ([call.method isEqualToString:@"getTestDeviceId"]) {
        result([[BaiduMobStat defaultStat] getTestDeviceId]);
    } else {
        result(FlutterMethodNotImplemented);
    }
}

// dart侧可选参数未传或传nil，oc侧接收到的是NSNull
- (id)validArgument:(id)argument {
    return [argument isEqual:[NSNull null]] ? nil : argument;
}

@end

File: android/src/main/kotlin/fl/baidu/mob/stat/BaiduMobStatPlugin.kt
package fl.baidu.mob.stat

import android.content.Context
import com.baidu.mobstat.StatService
import io.flutter.embedding.engine.plugins.FlutterPlugin
import io.flutter.plugin.common.MethodCall
import io.flutter.plugin.common.MethodChannel


class BaiduMobStatPlugin : FlutterPlugin, MethodChannel.MethodCallHandler {
    private var mContext: Context? = null
    private var channel: MethodChannel? = null

    override fun onAttachedToEngine(flutterPluginBinding: FlutterPlugin.FlutterPluginBinding) {
        channel = MethodChannel(flutterPluginBinding.binaryMessenger, "fl_baidu_mob_stat")
        channel!!.setMethodCallHandler(this)
        mContext = flutterPluginBinding.applicationContext

    }

    override fun onMethodCall(call: MethodCall, result: MethodChannel.Result) {
        when (call.method) {
            "setApiKey" -> {
                StatService.setAppKey(call.arguments as String)
                StatService.start(mContext)
                StatService.platformType(2)
                result.success(true)
            }
            "setAppChannel" -> {
                StatService.setAppChannel(mContext, call.arguments as String, true)
                result.success(true)
            }
            "setAppVersionName" -> {
                StatService.setAppVersionName(mContext, call.arguments as String)
                result.success(true)
            }
            "setDebug" -> {
                StatService.setDebugOn(call.arguments as Boolean)
                result.success(true)
            }
            "setDebugOn" -> {
                result.success(true)
            }
            "logEvent" -> {
                val eventId = call.argument<String>("eventId")
                val attributes = call.argument<MutableMap<String, String>>("attributes")
                StatService.onEvent(mContext, eventId, "", 1, attributes)
                result.success(true)
            }
            "logDurationEvent" -> {
                val eventId = call.argument<String>("eventId")
                val label = call.argument<String>("label")
                val duration = call.argument<Int>("duration")
                val attributes = call.argument<MutableMap<String, String>?>("attributes")
                if (duration != null) {
                    StatService.onEventDuration(mContext, eventId, label, duration.toLong(), attributes)
                }
                result.success(true)
            }
            "eventStart" -> {
                StatService.onEventStart(mContext, call.argument("eventId"), call.argument("label"))
                result.success(true)
            }
            "eventEnd" -> {
                val eventId = call.argument<String>("eventId")
                val label = call.argument<String>("label")
                val attributes = call.argument<MutableMap<String, String>?>("attributes")
                StatService.onEventEnd(mContext, eventId, label, attributes)
                result.success(true)
            }
            "pageStart" -> {
                StatService.onPageStart(mContext, call.arguments as String)
                result.success(true)
            }
            "pageEnd" -> {
                StatService.onPageEnd(mContext, call.arguments as String)
                result.success(true)
            }
            "getDeviceCuId" -> result.success("")
            "getTestDeviceId" -> result.success(StatService.getTestDeviceId(mContext))

            else -> {
                result.notImplemented()
            }
        }

    }

    override fun onDetachedFromEngine(binding: FlutterPlugin.FlutterPluginBinding) {
        channel!!.setMethodCallHandler(null)
    }

}